Create RemoveRedundantTSVRows.java

jwertz01 opened the pull request.

jwertz01 committed:
Create RemoveRedundantTSVRows.java

jjcarver commented on src/main/java/edu/ucsd/workflow/RemoveRedundantTSVRows.java:
```diff
@@ -0,0 +1,220 @@
+			String headerStr = input.readLine();
+			String[] header = headerStr.trim().split("\t");
+			String columnsStr = "";
+			Set<List<String>> uniqueLineIDs = new HashSet<>();
```
Your HashSet declaration is not properly typed. Type inference is supported in JRE 1.7 and later, but most of our code is still compliant only with 1.6. I guess this isn't really a problem, since we really ought to come out of our cave and start updating things for Java versions that have been out for almost a decade. Just something I noticed.

jjcarver commented on src/main/java/edu/ucsd/workflow/RemoveRedundantTSVRows.java:
```diff
@@ -0,0 +1,220 @@
+			}
+			for (int i=0; i<removeRedundantTSVRows.columns.length; i++) {
+				String column = removeRedundantTSVRows.columns[i];
+				if (!(Arrays.asList(header).contains(column))) {
```
It's super inefficient to cast the header array to a list in every iteration of this loop. The header will not change from one iteration to the next so this should be done statically before the loop starts.

jjcarver commented on src/main/java/edu/ucsd/workflow/RemoveRedundantTSVRows.java:
```diff
@@ -0,0 +1,220 @@
+				String[] line = lineStr.trim().split("\t");
+				List<String> uniqueLineID = new ArrayList<>();
+				for (int i=0; i<line.length; i++) {
+					if (Arrays.asList(removeRedundantTSVRows.columns).contains(
```
Same objection as above. This code will scale poorly to large files since you're repeating the same static operation on every line.

jjcarver commented on src/main/java/edu/ucsd/workflow/RemoveRedundantTSVRows.java:
```diff
@@ -0,0 +1,220 @@
+						uniqueLineID.add(line[i]);
+					}
+				}
+				if (!(uniqueLineIDs.contains(uniqueLineID))) {
```
I almost never use "!" in Java code. I find it's much more human readable to use " == false" in essentially every case. I would prefer we keep to these conventions as universally as possible in code projects maintained primarily by me.

jjcarver commented on src/main/java/edu/ucsd/workflow/RemoveRedundantTSVRows.java:
```diff
@@ -0,0 +1,220 @@
+					linesInOutputFile++;
+					output.println(lineStr);
+					uniqueLineIDs.add(uniqueLineID);
+				}
```
The logic here is good, but I don't quite understand the point of specifying a unique set of columns on the command line, and validating uniqueness against that. I guess the idea is that you're not just throwing out redundant lines, but really only validating against a particular set of columns? So for example, even if each row has 100 columns, you might want to filter out columns having duplicates of only columns A. B, and C? Similar to a SQL "group by" query. Because simply comparing line uniqueness with a set of strings is obviously much more efficient.

jwertz01 replied:
The idea is that, for example, there is a PSM file that has the exact same line twice, but with a different protein in the second line. If we don't care about the protein, we can reduce the size of the file by omitting the second line. The number and name of columns needs to be variable because, say, for one tool, we want one row per cluster index, whereas for another tool, we want one row per combination of filename and scan.

I thought about using concatenated strings instead of a list of strings to store the important columns in each line, but my issue with that was that could be weird edge cases such as
column A="AB", column B="CD" is equivalent to column A="A", column B="BCD".
The same problem can happen even with delimiters between the strings, unless there's some delimiter that will never occur in an input file.

jjcarver reviewed (changes requested):
The code is all correct and will work fine, but I have a few minor efficiency and code style suggestions. Also, I have a question about the overall logic just so I can understand it better. See line comments for details.

jwertz01 committed:
Addressing Jeremy's comments

Properly type declarations, cast header array to list outside of loop, use "== false" instead of "!"
